File: backend/app/routes/tasks.py
from typing import List

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from ..database import get_db
from .. import models, schemas
from ..auth import get_current_user

router = APIRouter()

@router.post("/", response_model=schemas.TaskResponse, status_code=status.HTTP_201_CREATED)
def create_task(
    task_in: schemas.TaskCreate,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user),
):
    """Create a new task in a sprint or backlog."""
    project = db.query(models.Project).filter(models.Project.id == task_in.project_id).first()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")

    if current_user.role.role_name.lower() != "manager" and current_user not in project.users:
         raise HTTPException(status_code=403, detail="Not authorized to create tasks for this project")
         
    new_task = models.Task(**task_in.model_dump())
    db.add(new_task)
    db.commit()
    db.refresh(new_task)
    return new_task

@router.put("/{task_id}/status", response_model=schemas.TaskResponse)
def update_task_status(
    task_id: int,
    status_update: schemas.TaskUpdateStatus,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user),
):
    """Update task status (drag and drop)."""
    task = db.query(models.Task).filter(models.Task.id == task_id).first()
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
        
    project = task.project
    if current_user.role.role_name.lower() != "manager" and current_user not in project.users:
         raise HTTPException(status_code=403, detail="Not authorized to update tasks for this project")
         
    task.status = status_update.status
    db.commit()
    db.refresh(task)
    return task

@router.get("/sprint/{sprint_id}", response_model=List[schemas.TaskResponse])
def get_tasks_for_sprint(
    sprint_id: int,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user),
):
    """Get all tasks for a specific sprint."""
    sprint = db.query(models.Sprint).filter(models.Sprint.id == sprint_id).first()
    if not sprint:
        raise HTTPException(status_code=404, detail="Sprint not found")
        
    project = sprint.project
    if current_user.role.role_name.lower() != "manager" and current_user not in project.users:
         raise HTTPException(status_code=403, detail="Not authorized to view tasks for this sprint")
         
    tasks = db.query(models.Task).filter(models.Task.sprint_id == sprint_id).all()
    return tasks

@router.get("/backlog/{project_id}", response_model=List[schemas.TaskResponse])
def get_backlog_tasks(
    project_id: int,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user),
):
    """Get all unassigned (backlog) tasks for a project."""
    project = db.query(models.Project).filter(models.Project.id == project_id).first()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
        
    if current_user.role.role_name.lower() != "manager" and current_user not in project.users:
         raise HTTPException(status_code=403, detail="Not authorized to view tasks for this project")
         
    tasks = db.query(models.Task).filter(models.Task.project_id == project_id, models.Task.sprint_id == None).all()
    return tasks

@router.put("/{task_id}", response_model=schemas.TaskResponse)
def update_task(
    task_id: int,
    task_update: schemas.TaskUpdate,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user),
):
    """Fully update a task (title, description, assignee, sprint_id, etc)."""
    task = db.query(models.Task).filter(models.Task.id == task_id).first()
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
        
    project = task.project
    if current_user.role.role_name.lower() != "manager" and current_user not in project.users:
         raise HTTPException(status_code=403, detail="Not authorized to update tasks for this project")
         
    update_data = task_update.model_dump(exclude_unset=True)
    for key, value in update_data.items():
        setattr(task, key, value)
        
    db.commit()
    db.refresh(task)
    return task

@router.delete("/{task_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_task(
    task_id: int,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user),
):
    """Delete a task."""
    task = db.query(models.Task).filter(models.Task.id == task_id).first()
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
        
    project = task.project
    if current_user.role.role_name.lower() != "manager" and current_user not in project.users:
         raise HTTPException(status_code=403, detail="Not authorized to delete tasks for this project")
         
    db.delete(task)
    db.commit()
    return None

# ==========================================
# COMMENTS
# ==========================================
@router.post("/{task_id}/comments", response_model=schemas.CommentResponse, status_code=status.HTTP_201_CREATED)
def create_comment(
    task_id: int,
    comment_in: schemas.CommentCreate,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user),
):
    task = db.query(models.Task).filter(models.Task.id == task_id).first()
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
        
    new_comment = models.Comment(**comment_in.model_dump(), task_id=task_id, author_id=current_user.id)
    db.add(new_comment)
    db.commit()
    db.refresh(new_comment)
    return new_comment

# ==========================================
# SUBTASKS
# ==========================================
@router.post("/{task_id}/subtasks", response_model=schemas.SubtaskResponse, status_code=status.HTTP_201_CREATED)
def create_subtask(
    task_id: int,
    subtask_in: schemas.SubtaskCreate,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user),
):
    task = db.query(models.Task).filter(models.Task.id == task_id).first()
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
        
    new_subtask = models.Subtask(**subtask_in.model_dump(), task_id=task_id)
    db.add(new_subtask)
    db.commit()
    db.refresh(new_subtask)
    return new_subtask

@router.put("/subtasks/{subtask_id}", response_model=schemas.SubtaskResponse)
def update_subtask(
    subtask_id: int,
    subtask_update: schemas.SubtaskUpdate,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user),
):
    subtask = db.query(models.Subtask).filter(models.Subtask.id == subtask_id).first()
    if not subtask:
        raise HTTPException(status_code=404, detail="Subtask not found")
        
    update_data = subtask_update.model_dump(exclude_unset=True)
    for key, value in update_data.items():
        setattr(subtask, key, value)
        
    db.commit()
    db.refresh(subtask)
    return subtask

@router.delete("/subtasks/{subtask_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_subtask(
    subtask_id: int,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user),
):
    subtask = db.query(models.Subtask).filter(models.Subtask.id == subtask_id).first()
    if not subtask:
        raise HTTPException(status_code=404, detail="Subtask not found")
        
    db.delete(subtask)
    db.commit()
    return None

File: backend/src/routes/tasks.ts
import { Router, Response } from 'express';
import prisma from '../lib/prisma';
import { authenticate, AuthRequest } from '../middleware/auth';

const router = Router();

/** Returns true if the user is a privileged role (Admin or Project Manager) */
function isPrivilegedUser(req: AuthRequest): boolean {
  const roleName: string = req.user?.role?.role_name ?? '';
  return roleName === 'Administrator' || roleName === 'Project Manager';
}

function parseId(param: string | string[] | undefined): number {
  return parseInt(Array.isArray(param) ? param[0] : (param ?? ''), 10);
}

// NOTE: Specific path routes must come before parameterized routes to avoid conflicts.

// GET /api/v1/tasks/sprint/:sprint_id — tasks for a sprint
router.get('/sprint/:sprint_id', authenticate, async (req: AuthRequest, res: Response): Promise<void> => {
  const sprint_id = parseId(req.params.sprint_id);
  if (isNaN(sprint_id)) {
    res.status(400).json({ detail: 'Invalid sprint ID' });
    return;
  }

  const sprint = await prisma.sprint.findUnique({
    where: { id: sprint_id },
    include: { project: { include: { users: true } } }
  });

  if (!sprint) {
    res.status(404).json({ detail: 'Sprint not found' });
    return;
  }

  const privileged = isPrivilegedUser(req);
  const isMember = sprint.project.users.some(u => u.id === req.user.id);

  if (!privileged && !isMember) {
    res.status(403).json({ detail: 'Not authorized to view tasks for this sprint' });
    return;
  }

  const tasks = await prisma.task.findMany({
    where: { sprint_id },
    include: {
      comments: { include: { author: true } },
      subtasks: true
    }
  });
  res.json(tasks);
});

// GET /api/v1/tasks/backlog/:project_id — backlog tasks (no sprint)
router.get('/backlog/:project_id', authenticate, async (req: AuthRequest, res: Response): Promise<void> => {
  const project_id = parseId(req.params.project_id);
  if (isNaN(project_id)) {
    res.status(400).json({ detail: 'Invalid project ID' });
    return;
  }

  const project = await prisma.project.findUnique({
    where: { id: project_id },
    include: { users: true }
  });

  if (!project) {
    res.status(404).json({ detail: 'Project not found' });
    return;
  }

  const privileged = isPrivilegedUser(req);
  const isMember = project.users.some(u => u.id === req.user.id);

  if (!privileged && !isMember) {
    res.status(403).json({ detail: 'Not authorized to view tasks for this project' });
    return;
  }

  const tasks = await prisma.task.findMany({
    where: { project_id, sprint_id: null },
    include: {
      comments: { include: { author: true } },
      subtasks: true
    }
  });
  res.json(tasks);
});

// PUT /api/v1/tasks/subtasks/:subtask_id — update subtask
router.put('/subtasks/:subtask_id', authenticate, async (req: AuthRequest, res: Response): Promise<void> => {
  const subtask_id = parseId(req.params.subtask_id);
  if (isNaN(subtask_id)) {
    res.status(400).json({ detail: 'Invalid subtask ID' });
    return;
  }

  const subtask = await prisma.subtask.findUnique({ where: { id: subtask_id } });
  if (!subtask) {
    res.status(404).json({ detail: 'Subtask not found' });
    return;
  }

  const updatedSubtask = await prisma.subtask.update({
    where: { id: subtask_id },
    data: req.body
  });
  res.json(updatedSubtask);
});

// DELETE /api/v1/tasks/subtasks/:subtask_id — delete subtask
router.delete('/subtasks/:subtask_id', authenticate, async (req: AuthRequest, res: Response): Promise<void> => {
  const subtask_id = parseId(req.params.subtask_id);
  if (isNaN(subtask_id)) {
    res.status(400).json({ detail: 'Invalid subtask ID' });
    return;
  }

  const subtask = await prisma.subtask.findUnique({ where: { id: subtask_id } });
  if (!subtask) {
    res.status(404).json({ detail: 'Subtask not found' });
    return;
  }

  await prisma.subtask.delete({ where: { id: subtask_id } });
  res.status(204).send();
});

// POST /api/v1/tasks/ — create task
router.post('/', authenticate, async (req: AuthRequest, res: Response): Promise<void> => {
  const { title, description, status, project_id, sprint_id, assignee_id, story_points, attachment_url } = req.body;

  if (!title || !project_id) {
    res.status(400).json({ detail: 'title and project_id are required' });
    return;
  }

  const project = await prisma.project.findUnique({
    where: { id: project_id },
    include: { users: true }
  });

  if (!project) {
    res.status(404).json({ detail: 'Project not found' });
    return;
  }

  const privileged = isPrivilegedUser(req);
  const isMember = project.users.some(u => u.id === req.user.id);

  if (!privileged && !isMember) {
    res.status(403).json({ detail: 'Not authorized to create tasks for this project' });
    return;
  }

  const new_task = await prisma.task.create({
    data: { title, description, status, project_id, sprint_id, assignee_id, story_points, attachment_url }
  });

  res.status(201).json(new_task);
});

// PUT /api/v1/tasks/:task_id/status — update task status only
router.put('/:task_id/status', authenticate, async (req: AuthRequest, res: Response): Promise<void> => {
  const task_id = parseId(req.params.task_id);
  if (isNaN(task_id)) {
    res.status(400).json({ detail: 'Invalid task ID' });
    return;
  }

  const { status } = req.body;

  const task = await prisma.task.findUnique({
    where: { id: task_id },
    include: { project: { include: { users: true } } }
  });

  if (!task) {
    res.status(404).json({ detail: 'Task not found' });
    return;
  }

  const privileged = isPrivilegedUser(req);
  const isMember = task.project.users.some(u => u.id === req.user.id);

  if (!privileged && !isMember) {
    res.status(403).json({ detail: 'Not authorized to update tasks for this project' });
    return;
  }

  const updatedTask = await prisma.task.update({
    where: { id: task_id },
    data: { status }
  });

  res.json(updatedTask);
});

// PUT /api/v1/tasks/:task_id — full task update
router.put('/:task_id', authenticate, async (req: AuthRequest, res: Response): Promise<void> => {
  const task_id = parseId(req.params.task_id);
  if (isNaN(task_id)) {
    res.status(400).json({ detail: 'Invalid task ID' });
    return;
  }

  const task = await prisma.task.findUnique({
    where: { id: task_id },
    include: { project: { include: { users: true } } }
  });

  if (!task) {
    res.status(404).json({ detail: 'Task not found' });
    return;
  }

  const privileged = isPrivilegedUser(req);
  const isMember = task.project.users.some(u => u.id === req.user.id);

  if (!privileged && !isMember) {
    res.status(403).json({ detail: 'Not authorized to update tasks for this project' });
    return;
  }

  const updatedTask = await prisma.task.update({
    where: { id: task_id },
    data: req.body
  });

  res.json(updatedTask);
});

// DELETE /api/v1/tasks/:task_id — delete task
router.delete('/:task_id', authenticate, async (req: AuthRequest, res: Response): Promise<void> => {
  const task_id = parseId(req.params.task_id);
  if (isNaN(task_id)) {
    res.status(400).json({ detail: 'Invalid task ID' });
    return;
  }

  const task = await prisma.task.findUnique({
    where: { id: task_id },
    include: { project: { include: { users: true } } }
  });

  if (!task) {
    res.status(404).json({ detail: 'Task not found' });
    return;
  }

  const privileged = isPrivilegedUser(req);
  const isMember = task.project.users.some(u => u.id === req.user.id);

  if (!privileged && !isMember) {
    res.status(403).json({ detail: 'Not authorized to delete tasks for this project' });
    return;
  }

  await prisma.task.delete({ where: { id: task_id } });
  res.status(204).send();
});

// POST /api/v1/tasks/:task_id/comments — add comment to task
router.post('/:task_id/comments', authenticate, async (req: AuthRequest, res: Response): Promise<void> => {
  const task_id = parseId(req.params.task_id);
  const { content } = req.body;

  const task = await prisma.task.findUnique({ where: { id: task_id } });
  if (!task) {
    res.status(404).json({ detail: 'Task not found' });
    return;
  }

  const comment = await prisma.comment.create({
    data: { content, task_id, author_id: req.user.id },
    include: { author: true }
  });
  res.status(201).json(comment);
});

// POST /api/v1/tasks/:task_id/subtasks — add subtask to task
router.post('/:task_id/subtasks', authenticate, async (req: AuthRequest, res: Response): Promise<void> => {
  const task_id = parseId(req.params.task_id);
  const { title, is_completed } = req.body;

  const task = await prisma.task.findUnique({ where: { id: task_id } });
  if (!task) {
    res.status(404).json({ detail: 'Task not found' });
    return;
  }

  const subtask = await prisma.subtask.create({
    data: { title, is_completed: is_completed ?? false, task_id }
  });
  res.status(201).json(subtask);
});

export default router;
